### PR TITLE
Lift compile-time dependencies from most libraries

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR is a large step towards a fully runtime-configured Mina node. This PR progressively refactors library-by-library to push the compile time configuration dependency up the stack.

Every commit in the history compiles, and makes a small amount of progress -- usually the requisite changes for a single library's disentanglement -- towards this goal.

Still unresolved is the question of how to set 'default' values. This PR takes no opinion on the correct solution there, and should be orthogonal to any solution there.

_Currently marked as draft, because I haven't broken the `Mina_lib` boundary quite yet._

### CODE REVIEW

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Comment by martyall** [comment]
!ci-build-me

**Review by georgeee — APPROVED** [review]
I'm not 100% confident that there are no more things like "Yikes! That's a nasty bug" in the PR.

But I'd say that at least when using node config based on `mainnet.mlh`, and not overriding constraint constants, genesis constants in runtime config, everything should work well.

And we're going to completely remove compile-time config very soon, so not a big concern.

PR is very clear in its objective, and reviewing it commit-by-commit is like reading a story.

**Review by dkijania — APPROVED** [review]

File: src/lib/cache_lib/impl.ml [hunk 1]
```diff
@@ -357,7 +372,7 @@ let%test_module "cache_lib test instance" =
                    any unconsumption handler calls" =
       setup () ;
       let logger = Logger.null () in
-      with_cache ~logger ~f:(fun cache ->
+      with_cache ~logger ~cache_exceptions:false ~f:(fun cache ->
```
Comment (volhovm) [inline src/lib/cache_lib/impl.ml:375]: why is `cache_exceptions` hardcoded to be `false` here and e.g. `true` in `catchup_scheduler`, while it seemed to be fully parameterized before ?

File: src/lib/ledger_catchup/normal_catchup.ml [hunk 2]
```diff
@@ -949,7 +949,7 @@ let%test_module "Ledger_catchup tests" =
       in
       let unprocessed_transition_cache =
         Transition_handler.Unprocessed_transition_cache.create ~logger
-          ~cache_exceptions:Node_config.cache_exceptions
+          ~cache_exceptions:true
```
Comment (volhovm) [inline src/lib/ledger_catchup/normal_catchup.ml:952]: Ah, because of unit tests. Ok.

File: src/app/rosetta/lib/account.ml [hunk 3]
```diff
@@ -30,11 +30,6 @@ end
 
 module Sql = struct
   module Balance_from_last_relevant_command = struct
-    let max_txns =
```
Comment (volhovm) [inline src/app/rosetta/lib/account.ml:33]: Was this just ... unused?

**Review by volhovm — APPROVED** [review]
So... Do we have a way to run the node and /check/ whether the concrete constants that are used in prod/tests are matching what they were before the change? Like as a debug output. That'd help not only with making sure this PR hasn't changed them, but also perhaps for diagnosing user's issues.

File: src/lib/node_config/unconfigurable_constants/node_config_unconfigurable_constants.ml [hunk 4]
```diff
@@ -0,0 +1,28 @@
+(* FIXME: These should be configurable. *)
```
[the remark this answers is not in the document]
Comment (georgeee) [reply src/lib/node_config/unconfigurable_constants/node_config_unconfigurable_constants.ml:1]: Yeah. Idea is that there were never configurable, but we want to make them configurable eventually. For now it makes sense to keep them in this module and get rid of them after figuring this out for constants that are already configurable.

**Comment by georgeee** [comment]
>The most complex thing about reviewing these changes is that it's really hard to trace which sets of constants get into which modules. 

Agree. I had the same concern. We need to select where to take a value, with the following options:
* runtime config
* compile-time config (`node_config`/`mina_compile_config`/...)
* unit-test-oriented compile-time config

When reviewing the PR I tried to ensure that unit test values are never used for non-unit-test code.

As of the distinction between compile-time and runtime configs, I'm a bit less worried. For `mainnet` we override in runtime config  only a couple of values, so compile-time/runtime configs mostly agree. Some environments might be broken, but this PR will be closely followed by a follow-up PR which removes compile time config completely, fully dropping this concern.

File: src/app/cli/src/cli_entrypoint/mina_cli_entrypoint.ml [hunk 5]
```diff
@@ -1673,18 +1673,7 @@ let snark_hashes =
...
-              (Lazy.force compiled).constraint_system_digests |> Lazy.force
-              |> List.map ~f:(fun (_constraint_system_id, digest) ->
-                     (* Throw away the constraint system ID to avoid changing the
-                        format of the output here.
-                     *)
-                     Md5.to_hex digest )
-          | None ->
-              []
-        in
+        let hashes = [] in
```
[the remark this answers is not in the document]
Comment (georgeee) [reply src/app/cli/src/cli_entrypoint/mina_cli_entrypoint.ml:1677]: I will fix it :)

File: src/app/rosetta/lib/account.ml (hunk 3 above)
Comment (georgeee) [reply src/app/rosetta/lib/account.ml:33]: Looks exactly this way. We shouldn't turn warnings off less freely, it seems so.

File: src/lib/cache_lib/impl.ml (hunk 1 above)
Comment (georgeee) [reply src/lib/cache_lib/impl.ml:375]: In unit tests we use only `dev` profile, and it was the only profile with `true` value. This rewriting changes behavior for `dev` profile, of course, but we weren't really using it for anything but unit tests :D

File: src/app/cli/src/cli_entrypoint/mina_cli_entrypoint.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (georgeee) [reply src/app/cli/src/cli_entrypoint/mina_cli_entrypoint.ml]: Done

**Comment by georgeee** [comment]
!ci-build-me

File: src/lib/mina_base/staged_ledger_hash.ml [hunk 6]
```diff
@@ -213,18 +213,12 @@ module Make_str (A : Wire_types.Concrete) = struct
...
-
     let typ : (var, value) Typ.t =
       Typ.transport (Typ.list ~length:length_in_bits Boolean.typ)
         ~there:(Fn.compose Fold.to_list fold) ~back:(fun _ ->
-          (* If we put a failwith here, we lose the ability to printf-inspect
-             * anything that uses staged-ledger-hashes from within Checked
-             * computations. It's useful when debugging to dump the protocol state
-             * and so we can just lie here instead. *)
-          warn_improper_transport () ; Lazy.force dummy )
+          (* TODO: We could just use a prover value. *)
```
[the remark this answers is not in the document]
Comment (mrmr1993) [reply src/lib/mina_base/staged_ledger_hash.ml:219]: We only use this for debug printing

**Comment by mrmr1993** [comment]
!ci-build-me

**Pull request merged**